features/cellular/framework/AT/AT_CellularNetwork.cpp

```diff
@@ -0,0 +1,905 @@
+        if (cid > cid_max) {
+            cid_max = cid;
+        }
+        char pdp_type_from_context[10];
+        int pdp_type_len = _at.read_string(pdp_type_from_context, sizeof(pdp_type_from_context) - 1);
+        if (pdp_type_len > 0) {
+            apn_len = _at.read_string(apn, sizeof(apn) - 1);
+            if (apn_len >= 0) {
+                if (strlen(_apn) && strcmp(apn, _apn) != 0 ) {
+                    continue;
+                }
+                nsapi_ip_stack_t pdp_stack = string_to_stack_type(pdp_type_from_context);
+                if (pdp_stack != DEFAULT_STACK) {
+                    if (get_modem_stack_type(pdp_stack)) {
+                        if (requested_stack == IPV4_STACK) {
+                            if (pdp_stack == IPV4_STACK || pdp_stack == IPV4V6_STACK) {
+                                _ip_stack_type = requested_stack;
+                                _cid = cid;
+                                break;
+                            }
+                        } else if (requested_stack == IPV6_STACK) {
+                            if (pdp_stack == IPV6_STACK || pdp_stack == IPV4V6_STACK) {
+                                _ip_stack_type = requested_stack;
+                                _cid = cid;
+                                break;
+                            }
+                        } else { // accept any but prefer to IPv6
+                            if (pdp_stack == IPV6_STACK || pdp_stack == IPV4V6_STACK) {
+                                _ip_stack_type = requested_stack;
+                                _cid = cid;
+                                break;
+                            }
+                            if (_ip_stack_type == DEFAULT_STACK) {
+                                _ip_stack_type = pdp_stack;
+                                _cid = cid;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    _at.resp_stop();
+    if (_cid == -1) { // no suitable context was found so create a new one
+        if (!set_new_context(requested_stack, cid_max+1)) {
+            return false;
+        }
+    }
+
+    // save the apn
+    if (apn_len > 0 && !strlen(_apn)) {
+        strncpy(_apn, apn, MAX_APN_LENGTH);
+    }
+
+    log_debug("Context id %d", _cid);
+    return true;
+}
+
+nsapi_ip_stack_t AT_CellularNetwork::string_to_stack_type(const char* pdp_type)
+{
+    nsapi_ip_stack_t stack = DEFAULT_STACK;
+    int len = strlen(pdp_type);
+
+    if (len == 6 && memcmp(pdp_type, "IPV4V6", len) == 0) {
+        stack = IPV4V6_STACK;
+    } else if (len == 4 && memcmp(pdp_type, "IPV6", len) == 0) {
+        stack = IPV6_STACK;
+    } else if (len == 2 && memcmp(pdp_type, "IP", len) == 0) {
+        stack = IPV4_STACK;
+    }
+    return stack;
+}
+
+nsapi_error_t AT_CellularNetwork::set_registration_urc(bool urc_on)
+{
+    RegistrationType reg_types[] = {C_EREG,        C_GREG,       C_REG};
```

Comment:
These should be a bit more const, and possibly static - `const RegistrationType` & `const char * const cmd_on`

Reply:
Added static const, commit
https://github.com/AriParkkila/mbed-os/commit/9ec4cb8e00dfe7719ed2742b2198ecf9c1c8b343